lib/api/apiUtils/object/abortMultipartUpload.js

```diff
@@ -105,9 +112,47 @@ function abortMultipartUpload(authInfo, bucketName, objectKey, uploadId, log,
                     }
                     cb();
                 });
+            }, () => next(null, mpuBucket, storedParts, destBucket, objectMD));
+        },
+        function deleteObjectMetadata(mpuBucket, storedParts, destBucket, objectMD, next) {
+            if (!objectMD || metadataValMPUparams.uploadId !== objectMD.uploadId) {
+                return next(null, mpuBucket, storedParts, destBucket, objectMD);
+            }
+            // In case there has been an error during cleanup after a complete MPU
+            // (e.g. failure to delete MPU MD in shadow bucket),
+            // we need to ensure that the MPU metadata is deleted.
+            log.info('Object has existing metadata, deleting them', {
+                method: 'abortMultipartUpload',
+                bucketName,
+                objectKey,
+            });
+            return metadata.deleteObjectMD(bucketName, objectKey, undefined, log, err => {
+                if (err) {
+                    log.error('error deleting object metadata', { error: err });
+                }
+                return next(err, mpuBucket, storedParts, destBucket, objectMD);
+            });
+        },
+        function deleteObjectData(mpuBucket, storedParts, destBucket, objectMD, next) {
+            if (!objectMD?.location) {
+                return next(null, mpuBucket, storedParts, destBucket);
+            }
+            // Filtering parts that has already been delete by the previous step
+            let partLocations = storedParts.map(item => item.value.partLocations);
+            partLocations = [].concat(...partLocations);
+            partLocations = new Set(partLocations.map(loc => loc.key));
```

Comment:
This looks under optimised: we map an array, then we deep copy it and recreate a set with another map. We should be able to merge it and create a set from a reduce of the partLocations keys